pkg/resourceinterpreter/configurableinterpreter/engine/luavm/lua.go

```diff
@@ -0,0 +1,239 @@
+	err := vm.setLib(l)
+	if err != nil {
+		return nil, err
+	}
+	// preload our 'safe' version of the OS library. Allows the 'local os = require("os")' to work
+	l.PreloadModule(lua.OsLibName, lifted.SafeOsLoader)
+
+	return &VM{LState: l}, err
+}
+
+// Load compile the script, return engine.Function of fnName int the scrip.
+// nRets indicates the number of returns of this function.
+func (vm *VM) Load(script string, fnName string, nRets int) (engine.Function, error) {
+	vm.lock.Lock()
+	defer vm.lock.Unlock()
+
+	err := vm.DoString(script)
+	if err != nil {
+		return nil, err
+	}
+
+	return engine.FunctionFunc(func(args ...interface{}) (rets []engine.Value, err error) {
+		vm.lock.Lock()
+		defer vm.lock.Unlock()
+
+		vArgs := make([]lua.LValue, len(args))
+		for i, arg := range args {
+			vArgs[i], err = toValue(vm.LState, arg)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		f := vm.GetGlobal(fnName)
+		if f.Type() == lua.LTNil {
+			return nil, fmt.Errorf("not found function %v", fnName)
+		}
+		if f.Type() != lua.LTFunction {
+			return nil, fmt.Errorf("not a function: %v", f.Type())
+		}
+
+		err = vm.CallByParam(lua.P{
+			Fn:      f,
+			NRet:    nRets,
+			Protect: true,
+		}, vArgs...)
+		if err != nil {
+			return nil, err
+		}
+
+		// get rets from stack: [ret1, ret2, ret3 ...]
+		rets = make([]engine.Value, nRets)
+		for i := range rets {
+			rets[i] = value{v: vm.Get(i + 1)}
+		}
+		vm.Pop(nRets)
+		return
+	}), nil
+}
+
+func (vm *VM) setLib(l *lua.LState) error {
+	for _, pair := range []struct {
+		n string
+		f lua.LGFunction
+	}{
+		{lua.LoadLibName, lua.OpenPackage},
+		{lua.BaseLibName, lua.OpenBase},
+		{lua.TabLibName, lua.OpenTable},
+		// load our 'safe' version of the OS library
+		{lua.OsLibName, lifted.OpenSafeOs},
+	} {
+		if err := l.CallByParam(lua.P{
+			Fn:      l.NewFunction(pair.f),
+			NRet:    0,
+			Protect: true,
+		}, lua.LString(pair.n)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type value struct {
+	v lua.LValue
+}
+
+var _ engine.Value = (*value)(nil)
+
+func (v value) IsNil() bool {
+	return v.v.Type() == lua.LTNil
+}
+
+func (v value) Int32() (int32, error) {
+	if v.v.Type() != lua.LTNumber {
+		return 0, fmt.Errorf("%#v is not number", v.v)
+	}
+	return int32(v.v.(lua.LNumber)), nil
+}
+
+func (v value) Bool() (bool, error) {
+	if v.v.Type() != lua.LTBool {
+		return false, fmt.Errorf("%#v is not bool", v.v)
+	}
+	return bool(v.v.(lua.LBool)), nil
+}
+
+func (v value) Into(obj interface{}) error {
+	t, err := conversion.EnforcePtr(obj)
+	if err != nil {
+		return fmt.Errorf("%#v is not pointer", v.v)
+	}
+
+	data, err := luajson.Encode(v.v)
+	if err != nil {
+		return fmt.Errorf("luajson Encode obj %#v error: %v", v.v, err)
+	}
+
+	// In lua, `{}` indicates array and table. Luajson encodes it to `[]`.
+	// So before converting to struct, convert it to '{}'.
+	// TODO: how to handle empty struct in fields, such as: {foo: {}}
```

Comment:
Does this make some functions unavailable?

Reply:
Yes, In @jameszhangyukun's demonstration, we saw the error log is caused by this qustion. 

For example, `GetReplicas` returns requirement with
```
{
    nodeClaim: {},
    resourceRequest: {
        cpu: 1000
    }
}
```
Then decode will error.
 
https://github.com/karmada-io/karmada/pull/2731#discussion_r1015424696

Reply:
Is there a solution to this problem? Some of the features are not available and I feel a little panicky.

Reply:
The master code has the same problem. So I will not fix it in the PR. I will try it in another PR..

Reply:
Ok, thx~